File: shared/protocol/envelope.py
"""Reference WABridge Protocol 1 envelope codec.

This module is intentionally dependency-free and is used as a protocol test
oracle. Native Windows and Android implementations must match these wire
invariants but must not copy Python into production.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

MAGIC = 0x5742
VERSION = 1
HEADER = struct.Struct("!HBBHHQI")
HEADER_SIZE = HEADER.size

CHANNEL_LIMITS = {
    1: 64 * 1024,
    2: 4 * 1024 * 1024,
    3: 1 * 1024 * 1024,
    4: 1 * 1024 * 1024,
    5: 256 * 1024,
}


class ProtocolError(ValueError):
    pass


@dataclass(frozen=True)
class Envelope:
    channel: int
    kind: int
    flags: int
    request_id: int
    payload: bytes


def encode(message: Envelope) -> bytes:
    if message.channel not in CHANNEL_LIMITS:
        raise ProtocolError("unknown channel")
    if not 0 <= message.kind <= 0xFFFF:
        raise ProtocolError("kind out of range")
    if not 0 <= message.flags <= 0xFFFF:
        raise ProtocolError("flags out of range")
    if message.request_id == 0:
        raise ProtocolError("request id must be non-zero")
    if len(message.payload) == 0 or len(message.payload) > CHANNEL_LIMITS[message.channel]:
        raise ProtocolError("payload exceeds channel limit")
    return HEADER.pack(
        MAGIC,
        VERSION,
        message.channel,
        message.kind,
        message.flags,
        message.request_id,
        len(message.payload),
    ) + message.payload


def decode(frame: bytes) -> Envelope:
    if len(frame) < HEADER_SIZE:
        raise ProtocolError("truncated header")
    magic, version, channel, kind, flags, request_id, length = HEADER.unpack_from(frame)
    if magic != MAGIC:
        raise ProtocolError("bad magic")
    if version != VERSION:
        raise ProtocolError("unsupported version")
    if channel not in CHANNEL_LIMITS:
        raise ProtocolError("unknown channel")
    if request_id == 0:
        raise ProtocolError("zero request id")
    if length == 0 or length > CHANNEL_LIMITS[channel]:
        raise ProtocolError("invalid payload length")
    if len(frame) != HEADER_SIZE + length:
        raise ProtocolError("frame length mismatch")
    return Envelope(channel, kind, flags, request_id, frame[HEADER_SIZE:])

File: android/app/src/main/java/com/wabridge/android/protocol/Envelope.kt
package com.wabridge.android.protocol

import java.nio.ByteBuffer
import java.nio.ByteOrder

class ProtocolException(message: String) : IllegalArgumentException(message)

data class Envelope(
    val channel: Int,
    val kind: Int,
    val flags: Int,
    val requestId: Long,
    val payload: ByteArray,
) {
    override fun equals(other: Any?): Boolean =
        other is Envelope && channel == other.channel && kind == other.kind &&
            flags == other.flags && requestId == other.requestId && payload.contentEquals(other.payload)

    override fun hashCode(): Int = 31 * (31 * (31 * (31 * channel + kind) + flags) + requestId.hashCode()) + payload.contentHashCode()
}

object EnvelopeCodec {
    private const val MAGIC: Short = 0x5742
    private const val VERSION: Byte = 1
    private const val HEADER_SIZE = 20
    private val channelLimits = mapOf(
        1 to 64 * 1024,
        2 to 4 * 1024 * 1024,
        3 to 1 * 1024 * 1024,
        4 to 1 * 1024 * 1024,
        5 to 256 * 1024,
    )

    fun encode(envelope: Envelope): ByteArray {
        val limit = channelLimits[envelope.channel] ?: throw ProtocolException("unknown channel")
        require(envelope.kind in 0..0xffff) { "kind out of range" }
        require(envelope.flags in 0..0xffff) { "flags out of range" }
        if (envelope.requestId == 0L) throw ProtocolException("zero request id")
        if (envelope.payload.isEmpty() || envelope.payload.size > limit) {
            throw ProtocolException("payload exceeds channel limit")
        }
        return ByteBuffer.allocate(HEADER_SIZE + envelope.payload.size)
            .order(ByteOrder.BIG_ENDIAN)
            .putShort(MAGIC)
            .put(VERSION)
            .put(envelope.channel.toByte())
            .putShort(envelope.kind.toShort())
            .putShort(envelope.flags.toShort())
            .putLong(envelope.requestId)
            .putInt(envelope.payload.size)
            .put(envelope.payload)
            .array()
    }

    fun decode(frame: ByteArray): Envelope {
        if (frame.size < HEADER_SIZE) throw ProtocolException("truncated header")
        val buffer = ByteBuffer.wrap(frame).order(ByteOrder.BIG_ENDIAN)
        if (buffer.short != MAGIC) throw ProtocolException("bad magic")
        if (buffer.get() != VERSION) throw ProtocolException("unsupported version")
        val channel = buffer.get().toInt() and 0xff
        val limit = channelLimits[channel] ?: throw ProtocolException("unknown channel")
        val kind = buffer.short.toInt() and 0xffff
        val flags = buffer.short.toInt() and 0xffff
        val requestId = buffer.long
        val length = buffer.int
        if (requestId == 0L) throw ProtocolException("zero request id")
        if (length <= 0 || length > limit) throw ProtocolException("invalid payload length")
        if (frame.size != HEADER_SIZE + length) throw ProtocolException("frame length mismatch")
        val payload = ByteArray(length)
        buffer.get(payload)
        return Envelope(channel, kind, flags, requestId, payload)
    }
}
